### PR TITLE
AP_Motors: Heli_Dual: refactor swash plate mixing

### PR DESCRIPTION
**Pull request by IamPete1**
This re-factors the dual heli swash plate mixing. The motors example was updated to allow setting of the `H_DUAL_MODE` in the heli test. This allowed a full sweep comparison or each dual heli type.

Results:
```
Dual_Heli H3 Longitudinal:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H3 Transverse:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H3 Intermeshing_or_Coaxial:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000

```

### CODE REVIEW

File: libraries/AP_Motors/AP_MotorsHeli_Dual.cpp [hunk 1]
```diff
@@ -302,81 +302,61 @@ void AP_MotorsHeli_Dual::calculate_scalars()
...
-            }
-        }
-    }
-    return swash_tilt;
+    // Differential cyclic roll is used for yaw and combined for roll
+    const float swash1_roll = roll_input + _yaw_scaler * yaw_input;
+    const float swash2_roll = roll_input - _yaw_scaler * yaw_input;
+
+    // cyclic is not used for pitch control
+    const float pitch = 0.0;
```
Comment (Gone4Dirt) [inline libraries/AP_Motors/AP_MotorsHeli_Dual.cpp:313]: nit pick: can we call this something like swash_pitch, just for readability.

File: libraries/AP_Motors/AP_MotorsHeli_Dual.cpp [hunk 2]
```diff
@@ -302,81 +302,61 @@ void AP_MotorsHeli_Dual::calculate_scalars()
...
+    // Calculate servo positions in swashplate library
+    _swashplate1.calculate(roll, swash1_pitch, swash1_coll);
+    _swashplate2.calculate(roll, swash2_pitch, swash2_coll);
+}
+
+// Mix and output swashplates for intermeshing
+void AP_MotorsHeli_Dual::mix_intermeshing(float pitch_input, float roll_input, float yaw_input, float collective1_input, float collective2_input)
+{
+    // Direct roll control on both swash plates
+    const float roll = roll_input;
```
Comment (Gone4Dirt) [inline libraries/AP_Motors/AP_MotorsHeli_Dual.cpp:347]: same nit pick: can we call this swash_roll

File: libraries/AP_Motors/AP_MotorsHeli_Dual.cpp [hunk 3]
```diff
@@ -516,17 +496,26 @@ void AP_MotorsHeli_Dual::move_actuators(float roll_out, float pitch_out, float c
...
+
+        case AP_MOTORS_HELI_DUAL_MODE_TRANSVERSE:
+            mix_transverse(pitch_out, roll_out, yaw_out, collective_out_scaled, collective2_out_scaled);
+            break;
+
+        case AP_MOTORS_HELI_DUAL_MODE_INTERMESHING:
+            mix_intermeshing(pitch_out, roll_out, yaw_out, collective_out_scaled, collective2_out_scaled);
+            break;
+
+        default:
```
Comment (Gone4Dirt) [inline libraries/AP_Motors/AP_MotorsHeli_Dual.cpp:513]: This is subtly different from the prior implementation where we effectively defaulted to tandem.
Might be worth adding a pre-arm on H_DUAL_MODE?  Just thinking of the user that has accidentally had it set to >3 and has previously had it working fine but now only has locked heads with nothing else to go on.  They will blame it on a firmware bug ....

File: libraries/AP_Motors/examples/AP_Motors_test/AP_Motors_test.cpp [hunk 4]
```diff
@@ -120,6 +120,15 @@ void setup()
...
                 }
 
+            } else if (strcmp(cmd,"dual_mode") == 0) {
+                if (frame_class != AP_Motors::MOTOR_FRAME_HELI_DUAL) {
+                    ::printf("dual_mode only supported by dual heli frame type (%i), got %i\n", AP_Motors::MOTOR_FRAME_HELI_DUAL, frame_class);
+                    exit(1);
+                }
+
+                // look away now, more dodgy param access.
+                AP_Int8 *dual_mode = (AP_Int8*)motors + AP_MotorsHeli_Dual::var_info[1].offset;
```
Comment (Gone4Dirt) [inline libraries/AP_Motors/examples/AP_Motors_test/AP_Motors_test.cpp:130]: How much work would it be to add a proper param accessor in AP_Motors_test?  Just thinking to make it easier to add other configs in the future?

File: libraries/AP_Motors/examples/AP_Motors_test/run_heli_comparison.py [hunk 5]
```diff
@@ -135,6 +148,7 @@ def run_sweep(frame_class, swash_type, dir_name):
     parser.add_argument("-s", "--swash-type", type=int, dest='swash_type', nargs="+", help="list of swashplate types to run comparison on. Defaults to test all types. Invalid for heli quad")
     parser.add_argument("-c", "--compare", action='store_true', help='Compare only, do not re-run tests')
     parser.add_argument("-p", "--plot", action='store_true', help='Plot comparison results')
+    parser.add_argument("-m", "--dual_mode", type=int, dest='dual_mode', help='Set DUAL_MODE, 0:Longitudinal, 1:Transverse, 2:Intermeshing/Coaxial')
```
Comment (Gone4Dirt) [inline libraries/AP_Motors/examples/AP_Motors_test/run_heli_comparison.py:151]:
```suggestion
    parser.add_argument("-m", "--dual_mode", type=int, dest='dual_mode', help='Set DUAL_MODE, 0:Longitudinal, 1:Transverse, 2:Intermeshing/Coaxial, default:run all')
```

Comment (Gone4Dirt) [reply libraries/AP_Motors/examples/AP_Motors_test/run_heli_comparison.py:151]: To help my rapidly dwindling memory :-D

**Review by Gone4Dirt — APPROVED** [review]
This reads a lot better IMHO, way easier to follow.
I have left a few minor comments which you can take or leave.

File: libraries/AP_Motors/AP_MotorsHeli_Dual.cpp (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsHeli_Dual.cpp]: done

File: libraries/AP_Motors/examples/AP_Motors_test/run_heli_comparison.py (hunk 5 above)
Comment (IamPete1) [reply libraries/AP_Motors/examples/AP_Motors_test/run_heli_comparison.py]: done

File: libraries/AP_Motors/AP_MotorsHeli_Dual.cpp (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsHeli_Dual.cpp]: done

**Review by bnsgeyer — APPROVED** [review]
@IamPete1 Thanks for cleaning this up.  I am not super picky about this type of stuff and I appreciate you and @Gone4Dirt spending the time cleaning up the code.  

I will add that one feature that will need to be added in the future for Tandem heli's is the ability to set a trim longitudinal cyclic input.  That way the cyclic pitch can be used to overcome drag and maintain speed without having high trim pitch attitudes.  I've had some partners asking about that. But that requires more than just changing the motors class

**Pull request merged**